File: client/src/app/services/game-play/sound.service.ts
/* eslint-disable no-console */
import { Injectable } from '@angular/core';
import { SettingsService } from '@app/services/divers/settings.service';
import { Speed } from '@common/enums/game-play/speed';
import { Sound } from '@common/enums/sound.enum';

@Injectable({
    providedIn: 'root',
})
export class SoundService {
    private playSpeed: Speed;
    private successSound: HTMLAudioElement;
    private errorSound: HTMLAudioElement;

    constructor(private readonly settingsService: SettingsService) {
        this.successSound = new Audio();
        this.errorSound = new Audio();
    }

    set speed(speed: Speed) {
        this.playSpeed = speed;
        this.successSound.playbackRate = speed;
        this.errorSound.playbackRate = speed;
    }

    playSuccess(): void {
        this.successSound = new Audio();
        this.successSound.volume = 0.2;
        this.successSound.src = this.settingsService.successSound;
        this.playSound(this.successSound);
    }

    playError(): void {
        this.errorSound = new Audio();
        this.errorSound.volume = 0.2;
        this.errorSound.src = this.settingsService.failureSound;
        this.playSound(this.errorSound);
    }

    pause() {
        this.successSound.pause();
        this.errorSound.pause();
    }

    resume() {
        if (this.isPlaying(this.successSound)) this.successSound.play().catch((e) => console.log('Error resuming success sound:', e));
        if (this.isPlaying(this.errorSound)) this.errorSound.play().catch((e) => console.error('Error resuming error sound:', e));
    }

    playSpecificSound = (sound: Sound): void => {
        let soundToPlay: HTMLAudioElement;

        if (sound.toString() === 'Yippee') {
            soundToPlay = new Audio('./assets/yippee.mp3');
        } else if (sound.toString() === 'Boiii') {
            soundToPlay = new Audio('./assets/boiii.mp3');
            setTimeout(() => {
                if (!soundToPlay.paused) {
                    soundToPlay.pause();
                    soundToPlay.currentTime = 0;
                }
                // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            }, 3000);
        } else {
            console.error(`No mp3 file found for: ${sound}`);
            return;
        }

        soundToPlay.volume = 0.2;
        this.playSound(soundToPlay);
    };

    private isPlaying(sound: HTMLAudioElement): boolean {
        return !sound.ended && sound.currentTime !== 0;
    }

    private playSound(sound: HTMLAudioElement): void {
        sound.currentTime = 0;
        sound.load();
        sound.playbackRate = this.playSpeed;
        sound.play().catch((error) => console.error('Error playing sound:', error));
    }
}

File: flutter_project/lib/service/sound-service.dart
import 'package:audioplayers/audioplayers.dart';
import 'package:flutter_project/common/enum.dart';
import 'package:flutter_project/service/settings-service.dart';

class SoundService {
  static final SoundService _instance = SoundService._privateConstructor();
  final AudioPlayer _successSoundPlayer = AudioPlayer();
  final AudioPlayer _errorSoundPlayer = AudioPlayer();
  final AudioPlayer audioPlayerObserver = AudioPlayer();

  final double _volume = 0.2;
  Speed speed = Speed.NORMAL;

  factory SoundService() {
    return _instance;
  }

  static SoundService get instance => _instance;

  SoundService._privateConstructor() {
    _initializePlayers();
  }

  void _initializePlayers() {
    _successSoundPlayer.setVolume(_volume);
    _errorSoundPlayer.setVolume(_volume);
    audioPlayerObserver.setVolume(_volume);
    _successSoundPlayer.setSource(AssetSource('success.mp3'));
    _errorSoundPlayer.setSource(AssetSource('error.wav'));
    speed = Speed.NORMAL;
  }

  void setSpeed(Speed newSpeed) {
    speed = newSpeed;
    _successSoundPlayer.setPlaybackRate(speed.value);
    _errorSoundPlayer.setPlaybackRate(speed.value);
  }

  Future<void> playSuccess() async {
    await _successSoundPlayer.stop();
    await _successSoundPlayer
        .play(AssetSource(SettingsService.instance.successSound));
  }

  Future<void> playError() async {
    await _errorSoundPlayer.stop();
    await _errorSoundPlayer
        .play(AssetSource(SettingsService.instance.failureSound));
  }

  Future<void> playSound(Sound sound) async {
    audioPlayerObserver.stop();
    final soundFileName = sound == Sound.Yippee ? 'yippee.mp3' : 'boiii.mp3';
    await audioPlayerObserver.play(AssetSource(soundFileName));
    Future.delayed(const Duration(seconds: 3), () {
      audioPlayerObserver.stop();
    });
  }

  void pause() {
    _successSoundPlayer.pause();
    _errorSoundPlayer.pause();
  }

  void resume() {
    _successSoundPlayer.resume();
    _errorSoundPlayer.resume();
  }
}
